### ISSUE TITLE
`RegexLiteral#to_s` returns invalid regex literal when literal value starts with whitespace character.

### ISSUE THREAD

**Issue by MakeNowJust** [crystal-lang/crystal#9016]
For example this code fails on compiling with `Error: Unterminated regular expression`.

```crystal
macro foo(x)
  p {{x}}
end

foo %r( )
```

Because `foo %r( )` expands to `p / /` and this first slash is recognized as divide operator and second is recognized as begin delimiter of regex.

**Comment by MakeNowJust** [comment]
Sorry. I now think #9017 fix is incomplete.
Please reopen this.

It changes regex literal value in such case. When I open #9017, it seems not so bad because it does not changes regex meaning. However this change causes some unexpected result:

```crystal
p  %r( ).source # => " "
p! %r( ).source # => "\\ "
```

The value should not change after macro expansion.

I have two ideas for fixing it:

  1. Escaping space character (`/\ /`) treats as special on parsing, like slash (`/\//`). For example `/\ /.source` returns `" "`. 
  2. Or, making `/` available as delimiter of `%r` regex literal (i.e. introducing `%r/.../` style regex literal) for explicit regex literal.

One of other ideas, using `%r(...)` regex literal, is incomplete too. For instance, `/ [)]/` is valid regex literal, and `%r( [)])` causes error. `%r{...}`, `%r[...]`, `%r<...>` and `%r|...|` have same problem.

I like the second opinion. It is clear and explicit. In fact, Rails code uses `%r/.../` at many places than `%r|...|`. (`rg '%r/' | wc -l` shows `40` and `rg '%r\|' | wc -l` shows `12`.)

What do you think?

/cc @asterite

**Comment by RX14** [comment]
Writing an escaping routine for `%r` which produced identical lexer ouput would be my first option, instead of adding syntax, but I don't mind either option.

**Comment by MakeNowJust** [comment]
@RX14 Re-using existed `%r` literal syntax is impossible or not so good, as the following reason:

- All delimiters except `<...>` have special meaning in regex syntax, so they cannot be escaped by backslash easily. If it is allowed (e.g. `\(` is recognized as `(` inside `%r(...)`), it causes double escaping problem.
- In `%r<...>` case, such a problem is not caused. However, unescaping `\<` and `\>` on the lexer is needed too. In short, it is almost same as the first idea of https://github.com/crystal-lang/crystal/issues/9016#issuecomment-612617893. In this view, I think specializing space escaping is more reasonable than specializing `\<` and `\>` escaping, because space is meaningful in Crystal syntax, so treating it on the parser is reasonable.

- - -

Another idea is ignoring delimiters inside character class `[...]` on the lexer. I think `%r(...)` with this treatment works well. But, implementation of this is too complex (at least adding a new field to `DelimiterState` is required). In addition, considering interpolation, I'm not sure it is non-problematic exactly.

One more idea is escaping by using interpolation `#{"..."}`. However this result looks wearied. For example, `#to_s` output for `/ [)]/` is `/#{" "}[)]/` or `/#{" [)]"}/` or `%r( [#{")"}])`. But, it is complete solution at least.

- - -

@asterite said at https://github.com/crystal-lang/crystal/pull/4429#issuecomment-302881705

> I think with just `|` we are done. I see no reason to support other delimiters. For example Elixir supports the same as Crystal. As always, copying Ruby is not what we should always do. Having more things to write a same concept is worse that having just a few ways (less things to learn). I also never saw something like `%*hello*` in Ruby, even though it's possible, it will only bring confusion.

However I believe `%r/.../` does not bring confusing, and we have reason to support this, because Crystal lacks explicit regex literal.

In fact Elixir support `~r/.../` style regex literal ([link](https://elixir-lang.org/getting-started/sigils.html)).

**Comment by MakeNowJust** [comment]
TL;DR of the above my comment:

  - Possible solutions other than `%r/.../` may introduce unexpected and nontrivial semantics to the language.
  - I think introducing `%r/.../` is the simplest and clearest solution of this problem.
  - And, adding `%r/.../` seems reasonable enough.

**Comment by asterite** [comment]
I think we just need to accept `\ ` as space.

**Comment by MakeNowJust** [comment]
@asterite It is acceptable. By the way, I want to know why `%r/.../` is not acceptable to you. One of reasons is consistency to other literals...?

**Comment by asterite** [comment]
Oh, `%r/.../` seems good to. We would just need to escape `/` using `\/`, right? Is that supported?

**Comment by MakeNowJust** [comment]
@asterite Yes, it is supported.

**Comment by asterite** [comment]
Cool! Then let's go with that solution then 👍 Thank you!

**Comment by MakeNowJust** [comment]
However `%/.../` and `%q/.../` should be valid?

**Comment by asterite** [comment]
Oh, wait... `%r/.../` isn't already a thing? I thought it was.

At this point, I don't know what's best to do :-)

**Comment by MakeNowJust** [comment]
Oh, sorry for confusing :pray:

**Comment by MakeNowJust** [comment]
Indeed I'd like to fix this bug. And, introducing `%r/.../` is not only a solution but also a new feature. Maybe it should be discussed in another issue...
